### DISCUSSION TITLE
Macro ScanCodeLexicalStateSwitch is referenced but not defined anywhere

### DISCUSSION THREAD

**Discussion by vsajip** [congo-cc/congo-parser-generator#86] [Q&A]
The `ScanCodeLexicalStateSwitch` macro isn't defined anywhere but is referenced, can I assume this usage is old and can be removed?
https://github.com/congo-cc/congo-parser-generator/blob/5effb1e0e6fde0b0af837cc422d750d0a1d7150c/src/templates/java/LookaheadRoutines.java.ftl#L333-L338

**Comment by revusky** [comment]
Yes, that's completely dead code. There isn't even a `ScanCodeLexicalStateSwitch` macro any more. A funny thing about that is that there is quite similar dead code in the ParserProductions template, i.e. here: https://github.com/congo-cc/congo-parser-generator/blob/main/src/templates/java/ParserProductions.java.ftl#L600

That is also vestigial stuff from an earlier implementation. That earlier implementation was wrong really because, really, these lexical changes have to be block oriented to really work properly. You have something like:

       LEXICAL_STATE SomeState (......)

or 

     ACTIVATE_TOKENS FOO (....) 

Basically, you enter the block, either parsing or in a lookahead, and when you leave it, the lexical state (including whatever tokens were active/inactive) gets restored to what it was before you entered. This is something that never really worked properly in the legacy JavaCC. The way it is dealt with in Congo now is with a macro with content here: https://github.com/congo-cc/congo-parser-generator/blob/main/src/templates/java/CommonUtils.java.ftl#L86

I'm actually quite proud of that code because the same macro handles both lexical state change and token activation/deactivation AND it handles it for both regular parsing and scanning ahead. Well, in any case, that is where the handling of these lexical state changes happens. The point where you are pointing in the code is just dead code that can be removed. In fact, there is no `LexicalStateSwitch` classname so it really can't enter that block, which is just as well, since there is no `ScanCodeLexicalStateSwitch` macro anyway. But, as you see, there is no mechanism to catch this in a FreeMarker template. In a compiled language, referring to a non-existent class or a non-existent subroutine would cause an error. In a FreeMarker template, this dead code can just continue to exist indefinitely, until somebody points it out. As you just did...
